styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -74,48 +76,79 @@
+                                 getShards(storage, counterId, executorService));
+    }
 
+    private static Map<Integer, Long> getShards(Storage storage, String counterId,
+                                                ExecutorService executorService) {
       Map<Integer, Long> fetchedShards = storage.shardsForCounter(counterId);
       if (fetchedShards.size() < NUM_SHARDS) {
         // The counter probably has not been initialized (so we have empty QueryResults). Also
         // possible that a prior initialize() crashed halfway, or we got a partial list of shards in
         // QueryResults due to eventual consistency. In any case, repeated initialization eventually
         // creates all NUM_SHARDS shards.
-        initialize(storage, counterId);
+        initialize(storage, counterId, executorService);
         fetchedShards = storage.shardsForCounter(counterId);
       }
-      this.counterId = counterId;
-      this.shards = fetchedShards;
+      return fetchedShards;
     }
 
     /**
      * Idempotent initialization, so that we don't reset an existing shard to zero - counterId may
      * have already been initialized and incremented by another process.
      */
-    private static void initialize(Storage storage, String counterId) {
-      for (int i = 0; i < NUM_SHARDS; i++) {
-        final int shardIndex = i;
-        try {
-          storage.runInTransaction(tx -> {
-            final Optional<Shard> shard = tx.shard(counterId, shardIndex);
+    private static void initialize(Storage storage, String counterId,
+                                   ExecutorService executorService) {
+      for (int i = 0; i < NUM_SHARDS; i += TRANSACTION_GROUP_SIZE) {
+        final int index = i;
+        executorService.execute(() -> initShard(storage, counterId, index, TRANSACTION_GROUP_SIZE));
+      }
+      try {
+        executorService.awaitTermination(20, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        LOG.warn("Error when initializing shards: ", e);
+      }
+    }
+
+    private static void initShard(Storage storage, String counterId, int index, int groupSize) {
+      try {
+        storage.runInTransaction(tx -> {
+          for (int i = 0; i < groupSize; i++) {
+            final Optional<Shard> shard = tx.shard(counterId, index + i);
             if (!shard.isPresent()) {
-              tx.store(Shard.create(counterId, shardIndex, 0));
+              tx.store(Shard.create(counterId, index + i, 0));
             }
-            return null;
-          });
-        } catch (IOException e) {
-          LOG.warn("Error when trying to create a shard entry in Datastore: ", e);
-        }
+          }
+          return null;
+        });
+      } catch (IOException e) {
+        LOG.warn("Error when trying to create a group of shards in Datastore: ", e);
       }
     }
+  }
+
+  private static class CounterSnapshot {
+
+    private static final int TRANSACTION_GROUP_SIZE = 15;
```

Comment:
Unused here. Where does 15 come from by the way?

Reply:
Ran a few tests and came to the conclusion that grouping 15 writes per transaction balances the overhead between thread creation and transaction creation/size

Reply:
This may change if we change how batch is done

Reply:
I've revised this, and see that I've done something wrong in the initial tests. the group/batch size doesn't seem to make a difference, and neither does parallelizing the transactions (due to the thread overhead and awaiting for thread shutdown). 

Simple batching has the highest impact.
That said, obviously, we do need to parallelize when we initialize shards on startup.